pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,492 @@
+	if oldMatches.Equal(newMatches) {
+		klog.V(2).InfoS("Processing Node UPDATE event, node cluster status not changed", "nodeName", node.Name)
+		return
+	}
+	affectedEIPs := oldMatches.Union(newMatches)
+	c.enqueueExternalIPPools(affectedEIPs)
+	klog.V(2).InfoS("Processed Node UPDATE event", "nodeName", node.Name, "affectedExternalIPPoolNum", affectedEIPs.Len())
+}
+
+func (c *Cluster) enqueueExternalIPPools(eips sets.String) {
+	for eip := range eips {
+		c.queue.Add(eip)
+	}
+}
+
+func (c *Cluster) enqueueExternalIPPool(obj interface{}) {
+	eip, ok := obj.(*v1alpha2.ExternalIPPool)
+	if !ok {
+		deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.ErrorS(decodingObjectErr, "Failed to process externalIPPool DELETE event", "obj", obj)
+			return
+		}
+		eip, ok = deletedState.Obj.(*v1alpha2.ExternalIPPool)
+		if !ok {
+			klog.ErrorS(decodingObjectTombstoneErr, "Failed to process externalIPPool DELETE event", "obj", deletedState.Obj)
+			return
+		}
+	}
+	c.queue.Add(eip.Name)
+}
+
+// newClusterMember get Node ip and return a cluster member(ip:clusterMemberlistPort) for joining
+func (c *Cluster) newClusterMember(node *corev1.Node) (string, error) {
+	nodeAddr, err := k8s.GetNodeAddr(node)
+	if err != nil {
+		klog.ErrorS(err, "Obtain IP address from K8s Node failed")
+		return "", err
+	}
+	member := fmt.Sprintf("%s:%d", nodeAddr, c.bindPort)
+	return member, nil
+}
+
+func (c *Cluster) allClusterMembers() (clusterNodes []string) {
+	nodes, err := c.nodeLister.List(labels.Everything())
+	if err != nil {
+		klog.ErrorS(err, "Listing Nodes failed")
+		return
+	}
+	klog.V(3).InfoS("List all cluster members", "node num", len(nodes))
+
+	for _, node := range nodes {
+		member, err := c.newClusterMember(node)
+		if err != nil {
+			continue
+		}
+		clusterNodes = append(clusterNodes, member)
+	}
+	return
+}
+
+func (c *Cluster) matchNodeSelector(node *corev1.Node, eip *v1alpha2.ExternalIPPool) bool {
+	nodeSelector, _ := metav1.LabelSelectorAsSelector(&eip.Spec.NodeSelector)
+	if nodeSelector.Matches(labels.Set(node.GetLabels())) {
+		return true
+	}
+	return false
+}
+
+func (c *Cluster) filterEIPsFromNodeLabels(node *corev1.Node) sets.String {
+	pools := sets.NewString()
+	eips, err := c.externalIPPoolLister.List(labels.Everything())
+	if err != nil {
+		klog.ErrorS(err, "Filter externalIPPools from nodeLabels failed")
+		return pools
+	}
+	for _, eip := range eips {
+		nodeSelector, _ := metav1.LabelSelectorAsSelector(&eip.Spec.NodeSelector)
+		if nodeSelector.Matches(labels.Set(node.GetLabels())) {
+			pools.Insert(eip.Name)
+		}
+	}
+	return pools
+}
+
+func (c *Cluster) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %s", controllerName)
+	defer klog.Infof("Shutting down %s", controllerName)
+
+	// In order to exit the cluster more gracefully , call Leave prior to shutting down.
+	defer close(c.nodeEventsCh)
```

Comment:
Again, can we move it to same place as defer c.queue.ShutDown() given they are same kind of cleanup operations? I don't see why we use the logs to separate them.

Reply:
done